### ISSUE TITLE
Some non-ASCII characters are misrepresented in the synopsis and title

### ISSUE THREAD

**Issue by martxw** [martxw/Humax_tvdiary#13]
Some non-ASCII characters, such as accented characters, apostrophes and pound signs, are shown as incorrect blobs. Interestingly, the same synopsis on SD and SH channels shows correctly and incorrectly.
Verify the character encoding and correct.

**Comment by martxw** [comment]
I experimented by changing tvdiary_utils.jim escape_json_string to add some ISO-6937 mappings. It works for some strings but not for others. That would be the SD/SH channel difference I noted above, although I'm not yet sure what indicates it. It should therefore actually be done closer to the DB access, as when it's JSON it's too late.

# Add the ISO-6937 accent prefix character. We lose the accents, so imperfect.
for {set i 0xC1} {$i <= 0xCF} {incr i} {
  set json_escape_map([format %c $i]) {}
}
# A more comprehensive solution would be to map every accent and letter pair to UTF-8 characters.
#set json_escape_map([format %c 0xC2]e) [format %c 0xC3][format %c 0xA9]
set json_escape_map([format %c 0xA3]) [format %c 0xC2][format %c 0xA3]

**Comment by hummypkg** [comment]
```humax# epg search 'The House that' | grep 'Name: The' | sed -n l
                          Name: The House That \243100k Built$
                          Name: The House That \302\243100k Built$
```
I need to find a solution for standard webif & RS too.

**Comment by hummypkg** [comment]
SD channel data is ISO-6937
HD channel data is UTF-8

I'm planning to update the EPG database to use UTF-8 throughout which should make it easier to fix this.

**Comment by hummypkg** [comment]
I've written a small character set conversion library that I'm going to link into the _epg_ utility to normalise data going into the database so that it's all UTF-8. That should simplify things for you!

https://github.com/hummypkg/libxconv

**Comment by hummypkg** [comment]
Updated EPG package now available, hopefully this will solve this issue.
(restart the EPG service after installation and allow time for the sql database to be updated)

**Comment by martxw** [comment]
Thanks! The EPG fix automatically fixes most of the TV Diary output.
To be recorded/Suggestions/EPG Search/Watchlist all get the text fresh from the EPG.
The legacy data used for Recorded/Watched/Monthly Programmes/Diary Search will potentially have ISO-6937 still. I verified this with an episode of "New: Building the Dream" recorded Tue 28 Oct 2014 22:00 and watched Wed 29 Oct 2014. The synopsis contains "fianc�ee". 
I can't find any example with the title containing an extended character, to test the Monthly Programme table.
The Inventory tab gets the text from the sidecar files, and I don't have an example containing extended characters.

I'll record "8mm 2" tonight as that has "fiancée" in the synopsis. I can then verify the Recorded & Watched history is now saved correctly, and that the Inventory is correct.

**Comment by hummypkg** [comment]
I'm intending to write a Jim module to do the character set conversion. Could be useful to fix legacy strings either as a one-off or as-needed.

**Comment by hummypkg** [comment]
Jim extension now available should you wish to use is:

```console
humax# opkg update
Downloading http://hpkg.tv/hdrfoxt2/base/Packages.gz.
...
humax# opkg install jim-xconv
Installing jim-xconv (1.0.0) to root...
Downloading http://hpkg.tv/hdrfoxt2/base/jim-xconv_1.0.0_mipsel.opk.
Configuring jim-xconv.

humax# jimsh
Welcome to Jim version 0.77
. require hexdump
. package require xconv
1.0
. set s "home made p\303at\302e"
home made pÃatÂe
. hexdump $s
0000: 68 6f 6d 65 20 6d 61 64 65 20 70 c3 61 74 c2 65  home made p.at.e

. set t [xconv $s]
home made pÃ¢tÃ©
. hexdump $t
0000: 68 6f 6d 65 20 6d 61 64 65 20 70 c3 a2 74 c3 a9  home made p..t..
```

**Comment by martxw** [comment]
Extended characters in the title cause some problems. Specifically finding the file as its name is based on the title. TV Diary logs:
Error getting details of humaxtv file /mnt/hd2/My Video/Handbags Under \xc2\xa3100_20170311_2300.ts. Ignoring it. [could not read "/mnt/hd2/My Video/Handbags Under \xc2\xa3100_20170311_2300.ts": No such file or directory]
... and it fails to show the programme as being recorded.

The WebIF dropdown status shows: "Playing Handbags Under \xc2\xa3100_20170311_2300".
Browsing in WebIF shows the file as "Handbags Under £100_20170311_2300.ts (0 bytes)".

The TV Diary inventory correctly shows the title, but the synopsis has a blob character.

When trying to watch the programme, the Humax TV UI says the the recording failed - the programme appears not to have been broadcast.

In Telnet, ls shows:
humax# ls -l Handbags*
-rw-r--r--    1 root     root          2072 Mar 12 00:23 Handbags Under ??100_20170311_2300.hmt
-rw-r--r--    1 root     root             0 Mar 11 23:00 Handbags Under ??100_20170311_2300.nts
-rw-r--r--    1 root     root             0 Mar 11 23:00 Handbags Under ??100_20170311_2300.ts

I'm not sure what to make of the failure to record. The WebUI and TV Diary saw the file open for a good length of time, and the red ring showed it was recording.
As far as TV Diary goes, there's still some work to do, and more testing is necessary.

**Comment by martxw** [comment]
Recorded "The House That £100k Built", on BBC 2 & BBC 2 HD. Both show the files growing OK while recording.
The TV Diary log reports failure to access the files, and so nothing shows as recording.
After 15 minutes I stop the SD and try watching that live, so it should be accessing the /mnt/hd2/Tsr/0.ts file rather than the recorded one with extended characters.
This works - the log shows it's watching live, and the characters seem correct. Ditto when I stop the HD & watch live.
tvdiary_status is using lsof, and that reports the file path as "The House That \xc2\xa3100k Built_20170312_1623.ts". The file path includes the backslashes, so to evaluate them as the extended characters I need to call the subst function.
In tvdiary_status get_data, change:
  set filepath [regsub {^[^/]*} $line {}]
to:
  set filepath [subst [regsub {^[^/]*} $line {}]]

**Comment by martxw** [comment]
I think there's just legacy data in the DB to sort out.
EG the synopsis table:
11594	Jason & Riikka: Following people trying to build a house on less than \xa3100k. Jason and his Finnish wife Riikka don't want a big house, but one which perfectly fits their needs. Also in HD. [S]
11595	Jason & Riikka: Following people trying to build a house on less than \xc2\xa3100k. Jason and his Finnish wife Riikka don't want a big house, but one which perfectly fits their needs. [HD] [S]
The first is ISO-6937, and the second is UTF-8.

This can be fixed by calling xconv before every JSON string is output, but this is wasteful.

> package require xconv
> proc escape_json_string { in } {
>   return [string map $::json_escape_map [xconv $in]]
> }

An alternative would be to have a one-off process which updates the database. But this would require locking the rest of TV Diary out for a period of time.
I'll experiment to see how long it takes across my DB that goes back to Nov 2013.

**Comment by hummypkg** [comment]
If you want a separate data point:

```console
humax# time sqlite3 -cmd '.load /mod/lib/sql/libqxconv.so' /mod/etc/tvdiary.db '
update synopses set synopsis = xconv(synopsis);'
real    0m 11.84s
user    0m 3.85s
sys     0m 1.32s
humax# sqlite3 /mod/etc/tvdiary.db 'select count(*) from synopses'
8866
```

**Comment by hummypkg** [comment]
NB: The SQLite3 plugin file is going to be renamed to _/mod/lib/sql/xconv.so_ in a package update this evening but its functionality won't change.

**Comment by martxw** [comment]
Oops. subst is dangerous! It evaluates the string.

> . set d "DANGER"
> . DANGER
> . set f "\[\$d\]/The House That \\xc2\\xa3100k Built_20170312_1623.ts"
[$d]/The House That \xc2\xa3100k Built_20170312_1623.ts
> . subst $f
> invalid command name "DANGER"

I noticed this because the tvdiary.log file was reporting that the commands Saved and Deleted don't exist, and these correspond to folders I have named [Saved] and[Deleted].
Command, Variable and Backslash substitutions occur, but I only want backslash substitutions. Luckilly subst has switches for that.

> . subst -nocommands -novariables $f
> [$d]/The House That ┬ú100k Built_20170312_1623.ts
